### ISSUE TITLE
Invalid RID

### ISSUE THREAD

**Issue by zaksnet** [godotengine/godot#33797]
**Godot version:**
Beta 1
**OS/device including version:**
Windows 10 1903
**Issue description:**
No matter which function i use, the RID object always gives me `Invalid RID`.
**Steps to reproduce:**
Try any of these:
`var ps_rid = get_viewport().get_world().get_space() # ps_rid is Invalid RID`
`var vs_rid = get_viewport().get_world().get_canvas() # vs_rid is Invalid RID`
```
var rigidBody = RigidBody2D.new()
var rb = rigidBody.instance()
add_child(rb)
print(rb.get_rid()) # prints Invalid RID
```

**Comment by bojidar-bg** [comment]
The only mention of `Invalid RID` in the code is in the Inspector (and mostly makes sense, as the RID is valid in the debugged game, not in the editor):
https://github.com/godotengine/godot/blob/37b230fe3af3c7e20a6f938d03b61a79b117d354/editor/editor_properties.cpp#L2060-L2068

I don't think the `print` statement can result in `Invalid RID`, it should print `[RID]` from what I can see.

**Comment by zaksnet** [comment]
Alright, that makes sense, BUT, trying to call any kind of method on that RID object just gives me:
![image](https://user-images.githubusercontent.com/1850856/69366558-48447200-0c9f-11ea-953e-fe91accece24.png)
The image is from:
```
	var vs_rid = get_viewport().get_world_2d().get_canvas()
	var ci_rid = vs_rid.canvas_item_create()
```
I havent been able to call any kind of method on any returned RID object.

**Comment by zaksnet** [comment]
Alright, that makes sense, BUT, trying to call any kind of method on that RID object just gives me:
![image](https://user-images.githubusercontent.com/1850856/69366558-48447200-0c9f-11ea-953e-fe91accece24.png)
The image is from:
```
	var vs_rid = get_viewport().get_world_2d().get_canvas()
	var ci_rid = vs_rid.canvas_item_create()
```
I havent been able to call any kind of method on any returned RID object.

**Comment by zaksnet** [comment]
> 

> The only mention of `Invalid RID` in the code is in the Inspector (and mostly makes sense, as the RID is valid in the debugged game, not in the editor):
> https://github.com/godotengine/godot/blob/37b230fe3af3c7e20a6f938d03b61a79b117d354/editor/editor_properties.cpp#L2060-L2068
> 
> I don't think the `print` statement can result in `Invalid RID`, it should print `[RID]` from what I can see.

NVM, i think my confuzion was just with the `Invalid RID`. using `VisualServer.canvas_item_create()` with the visual server `singleton` instead of the `RID` seems to work

**Comment by zaksnet** [comment]
Looking back, the problem started when i was following the [docs ](https://docs.godotengine.org/en/latest/tutorials/optimization/using_servers.html) in an attempt to replicate:
```
func _body_moved(state, index):
    # Created your own canvas item, use it here.
    VisualServer.canvas_item_set_transform(canvas_item, state.transform)

func _ready():
    # Create the body.
    var body = Physics2DServer.body_create()
    Physics2DServer.body_set_mode(body, Physics2DServer.BODY_MODE_RIGID)
    # Add a shape.
    var shape = RectangleShape2D.new()
    shape.extents = Vector2(10, 10)
    # Make sure to keep the shape reference!
    Physics2DServer.body_add_shape(body, shape)
    # Set space, so it collides in the same space as current scene.
    Physics2DServer.body_set_space(body, get_world_2d().space)
    # Move initial position.
    Physics2DServer.body_set_state(body, Physics2DServer.BODY_STATE_TRANSFORM, Transform2D(0, Vector2(10, 20)))
    # Add the transform callback, when body moves
    # The last parameter is optional, can be used as index
    # if you have many bodies and a single callback.
    Physics2DServer.body_set_force_integration_callback(body, self, "_body_moved", 0)
```
` Physics2DServer.body_add_shape(body, shape)` That like is wrong. shape variable is not an RID. You have to get the RID first using `get_rid()`. But then i saw the `Invalid RID` in the debugger and i assumed that was the error :dizzy_face:.
So to sum up, first there needs to be a fix for the `Invalid RID` message and second for the docs. I will make the changes myself later with a PR. Thanks @bojidar-bg :punch:

**Comment by Zylann** [comment]
> Physics2DServer.body_add_shape(body, shape) That line is wrong. shape variable is not an RID. You have to get the RID first using get_rid()

Is Godot actually converting into a RID if the parameter inherits `Resource`?

**Comment by zaksnet** [comment]
> > Physics2DServer.body_add_shape(body, shape) That line is wrong. shape variable is not an RID. You have to get the RID first using get_rid()
> 
> Isn't Godot actually converting into a RID if the parameter inherits `Resource`? That may add to the confusion but I think that's what happens.

@Zylann Yeah, i think you're right 😵😵. Now I'm lost.

**Comment by ThereIs0Spoon** [comment]
This might be just restating what has been already said, but I see an identical behaviour on the following:

var D:DynamicFont
var R:RID
var i:int

....
R = $Sprite.get_rid()
i = R.get_id()
print(R)
print(i)
....
D.size = 24
D.draw(R, Vector2(0.0, 0.0), "abc123")

In the output window I get:
[RID]
151
and get the following error from DynamicFont.draw()
![immagine](https://user-images.githubusercontent.com/60713866/73967486-4726f680-4918-11ea-9fd0-4c1fd39704eb.png)

**Comment by Xrayez** [comment]
Furthermore, if you somehow expose RID property, it won't be saved properly in a `tscn`, which will lead to parse error on opening. I've only stumbled upon this while accidentally exposing an `RID` as a property using C++ with some combination of property hints and usage. Exporting RID via GDScript and saving works alright.

**Comment by Listwon** [comment]
I'm trying to use `Physics2DServer.area_set_area_monitor_callback` and with breakpoint inside callback function I always get Invalid RID in inspector view.
![Invalid RID](https://user-images.githubusercontent.com/1554127/93218741-d88dff00-f76a-11ea-8991-853fc52d6e8d.png)

Sample project in Godot 3.2.2
[bullet-shower.zip](https://github.com/godotengine/godot/files/5225279/bullet-shower.zip)

**Comment by FrederickDesimpel** [comment]
Creating and working with RIDs directly using VisualServer works, so that Invalid RID message in the inspector is indeed misleading.
